File: elixir-impl/test/controller_test.exs
defmodule JobProcessor.ControllerTest do

  use ExUnit.Case, async: true
  alias JobProcessor.Controller

  @tasks_list [
    %{"command" => "touch /tmp/file1", "name" => "task-1"},
    %{"command" => "cat /tmp/file1", "name" => "task-2", "requires" => ["task-3"]},
    %{"command" => "echo 'Hello World!' > /tmp/file1", "name" => "task-3", "requires" => ["task-1"]},
    %{"command" => "rm /tmp/file1", "name" => "task-4", "requires" => ["task-2", "task-3"]}
  ]

  test "Controller.process_tasks – empty input returns error" do
    assert Controller.process_tasks(%{}) == Controller.generic_error_response()
  end

  test "Controller.process_tasks – wrong input returns error" do
    assert Controller.process_tasks(%{ "any" => 1, "other" => [] }) == Controller.generic_error_response()
  end

  test "Controller.process_tasks – empty tasks list is OK" do
    assert Controller.process_tasks(%{ "tasks" => [] }) == {200, %{"tasks" => []}, :json}
  end

  test "Controller.process_tasks – wrong response_type returns JSON" do
    assert Controller.process_tasks(%{ "tasks" => [], "response_type" => "xml" }) == {200, %{"tasks" => []}, :json}
  end

  test "Controller.process_tasks – bash response type returns text" do
    {status, body, content_type} = Controller.process_tasks(%{ "tasks" => [], "response_type" => "bash" })
    assert status == 200 and is_binary(body) and content_type == :text
  end

  test "Controller.process_tasks – tasks are properly ordered in bash response" do
    {_, body, _} = Controller.process_tasks(%{ "response_type" => "bash", "tasks" => @tasks_list })
    assert body == "#!/usr/bin/env bash\ntouch /tmp/file1\necho 'Hello World!' > /tmp/file1\ncat /tmp/file1\nrm /tmp/file1" 
  end

  test "Controller.process_tasks – tasks are properly ordered in JSON response" do
    {_, body, _} = Controller.process_tasks(%{ "tasks" => @tasks_list })
    assert %{ "tasks" => [
      %{"name" => "task-1" },
      %{"name" => "task-3" },
      %{"name" => "task-2" },
      %{"name" => "task-4" }
    ]} = body
  end

  test "Controller.process_tasks – JSON response only includes command and name for each task" do
    {_, body, _} = Controller.process_tasks(%{ "tasks" => @tasks_list })
    valid_keys = MapSet.new(["name", "command"])
    assert Enum.all?(body["tasks"], fn task -> 
      MapSet.new(Map.keys(task)) == valid_keys
    end)
  end

end

File: erlang-impl/test/controller_test.erl
-module(controller_test).
-include_lib("eunit/include/eunit.hrl").

-define(TASKS_LIST, [
    #{<<"command">> => <<"touch /tmp/file1">>, <<"name">> => <<"task-1">>},
    #{<<"command">> => <<"cat /tmp/file1">>, <<"name">> => <<"task-2">>, <<"requires">> => [<<"task-3">>]},
    #{<<"command">> => <<"echo 'Hello World!' > /tmp/file1">>, <<"name">> => <<"task-3">>, <<"requires">> => [<<"task-1">>]},
    #{<<"command">> => <<"rm /tmp/file1">>, <<"name">> => <<"task-4">>, <<"requires">> => [<<"task-2">>, <<"task-3">>]}
]).

sort_test_() ->
  [
    {
      "controller:process_tasks – empty input returns error",
      fun() ->
        ?assertEqual(
          controller:generic_error_response(), 
          controller:process_tasks(#{})
        )
      end
    },
    {
      "controller:process_tasks – wrong input returns error",
      fun() ->
        ?assertEqual(
          controller:generic_error_response(), 
          controller:process_tasks(#{ <<"any">> => 1, <<"other">> => [] })
        )
      end
    },
    {
      "controller:process_tasks – empty tasks list is OK",
      fun() ->
        ?assertEqual(
          {200, #{<<"tasks">> => []}, json}, 
          controller:process_tasks(#{ <<"tasks">> => [] })
        )
      end
    },
    {
      "controller:process_tasks – wrong response_type returns JSON",
      fun() ->
        ?assertEqual(
          {200, #{<<"tasks">> => []}, json}, 
          controller:process_tasks(
            #{ <<"tasks">> => [], <<"response_type">> => <<"xml">> }
          )
        )
      end
    },
    {
      "controller:process_tasks – bash response type returns text",
      fun() ->
        {Status, Body, ContentType} = controller:process_tasks(
          #{ <<"tasks">> => [], <<"response_type">> => <<"bash">>}
        ),
        ?assertEqual(200, Status),
        ?assert(is_binary(Body)),
        ?assertEqual(text, ContentType)
      end
    },
    {
      "controller:process_tasks – tasks are properly ordered in bash response",
      fun() ->
        {_, Body, _} = controller:process_tasks(
          #{ <<"response_type">> => <<"bash">>, <<"tasks">> => ?TASKS_LIST }
        ),
        ?assertEqual(
          <<"#!/usr/bin/env bash\ntouch /tmp/file1\necho 'Hello World!' > /tmp/file1\ncat /tmp/file1\nrm /tmp/file1">>, 
          Body
        )
      end
    },
    {
      "controller:process_tasks – tasks are properly ordered in JSON response",
      fun() ->
        {_, Body, _} = controller:process_tasks(#{ <<"tasks">> => ?TASKS_LIST }),
        ?assertEqual(#{ <<"tasks">> => [
          #{<<"command">> => <<"touch /tmp/file1">>, <<"name">> => <<"task-1">>},
          #{<<"command">> => <<"echo 'Hello World!' > /tmp/file1">>, <<"name">> => <<"task-3">>},
          #{<<"command">> => <<"cat /tmp/file1">>, <<"name">> => <<"task-2">>},
          #{<<"command">> => <<"rm /tmp/file1">>, <<"name">> => <<"task-4">>}
        ]}, Body)
      end
    },
    {
      "controller:process_tasks – JSON response only includes command and name for each task",
      fun() ->
        {_, Body, _} = controller:process_tasks(#{<<"tasks">> => ?TASKS_LIST}),
        lists:all(fun(Task) ->
          maps:size(Task) =:= 2 andalso
          maps:is_key(<<"name">>, Task) andalso
          maps:is_key(<<"command">>, Task)
        end, maps:get(<<"tasks">>, Body))
      end
    }
  ].
